Firestore/Source/API/FSTUserDataReader.mm

```diff
@@ -277,51 +318,70 @@ - (FieldValue)parsedQueryValue:(id)input allowArrays:(bool)allowArrays {
       if (context.array_element() && context.data_source() != UserDataSource::ArrayArgument) {
         ThrowInvalidArgument("Nested arrays are not supported");
       }
-      return [self parseArray:(NSArray *)input context:std::move(context)];
+      return absl::optional<google_firestore_v1_Value>([self parseArray:(NSArray *)input
+                                                                context:std::move(context)]);
     } else {
-      return [self parseScalarValue:input context:std::move(context)];
+      return absl::optional<google_firestore_v1_Value>([self parseScalarValue:input
+                                                                      context:std::move(context)]);
     }
   }
 }
 
-- (FieldValue)parseDictionary:(NSDictionary<NSString *, id> *)dict
-                      context:(ParseContext &&)context {
+- (google_firestore_v1_Value)parseDictionary:(NSDictionary<NSString *, id> *)dict
+                                     context:(ParseContext &&)context {
+  google_firestore_v1_Value result;
```

Comment:
Nit: consider initializing `result` using empty curly braces (I think the code initializes all the values correctly, but it just feels like a safer default).

Reply:
Done